### PR TITLE
CB-18190 This saves Blueprint Upgrade Options from custom builds in the database.

### PR DESCRIPTION
**Pull request by vivprabu**
This saves Blueprint Upgrade Options from custom builds in the database.

### CODE REVIEW

**Comment by keyki** [comment]
I think the problem is that custom blueprints does not have this field. Customers don't know that they should use it as it's not even documented. Instead of forcing them to set a value to it - which is entirely on them what is the value, we should ignore what the value is in case of custom blueprints. This means if it's custom we treat it as it would be set to GA.

**Comment by gergopapi2** [comment]
> I think the problem is that custom blueprints does not have this field. Customers don't know that they should use it as it's not even documented. Instead of forcing them to set a value to it - which is entirely on them what is the value, we should ignore what the value is in case of custom blueprints. This means if it's custom we treat it as it would be set to GA.

@keyki I see. The customer actually did set the blueprintUpgrdaeOption field to true, but there was a bug in our code and the field was not saved properly. 
If, however, the requirements have changed, or we think the problem should be solved in a different way, then please let's discuss.

**Comment by gergopapi2** [comment]
> > I think the problem is that custom blueprints does not have this field. Customers don't know that they should use it as it's not even documented. Instead of forcing them to set a value to it - which is entirely on them what is the value, we should ignore what the value is in case of custom blueprints. This means if it's custom we treat it as it would be set to GA.
> 
> @keyki I see. The customer actually did set the blueprintUpgrdaeOption field to true, but there was a bug in our code and the field was not saved properly. If, however, the requirements have changed, or we think the problem should be solved in a different way, then please let's discuss.

We discussed with @keyki and it would be better to change the requirements for custom blueprints. I will modify the jira shortly. @vivprabu let's then discuss.

File: core/src/test/java/com/sequenceiq/cloudbreak/converter/v4/blueprint/BlueprintV4RequestToBlueprintConverterTest.java
Comment (lajosrodek) [inline core/src/test/java/com/sequenceiq/cloudbreak/converter/v4/blueprint/BlueprintV4RequestToBlueprintConverterTest.java:6]: Could you please migrate this class to JUnit 5?

File: template-manager-cmtemplate/src/main/java/com/sequenceiq/cloudbreak/cmtemplate/utils/BlueprintUtils.java
Comment (lajosrodek) [inline template-manager-cmtemplate/src/main/java/com/sequenceiq/cloudbreak/cmtemplate/utils/BlueprintUtils.java:144]: Could you please cover with UT? Be sure to use JUnit 5, thanks.

**Review by lajosrodek — CHANGES_REQUESTED** [review]
The logic seems alright, though I suggest waiting for an approval from Duck as well.

A general comment: The commit msg has an empty body, so zero description of the changes. Please add some details, at least the PR description as a bare minimum, because it makes easier to determine the purpose later when looking at the Git history.

File: core/src/test/java/com/sequenceiq/cloudbreak/converter/v4/blueprint/BlueprintV4RequestToBlueprintConverterTest.java
Comment (vivprabu) [reply core/src/test/java/com/sequenceiq/cloudbreak/converter/v4/blueprint/BlueprintV4RequestToBlueprintConverterTest.java:6]: done

File: template-manager-cmtemplate/src/main/java/com/sequenceiq/cloudbreak/cmtemplate/utils/BlueprintUtils.java
Comment (vivprabu) [reply template-manager-cmtemplate/src/main/java/com/sequenceiq/cloudbreak/cmtemplate/utils/BlueprintUtils.java:144]: This is done in line 124 and 134 in the test class, Lajos.

Comment (lajosrodek) [reply template-manager-cmtemplate/src/main/java/com/sequenceiq/cloudbreak/cmtemplate/utils/BlueprintUtils.java:144]: Sorry, which test class? I cannot see `BlueprintUtilsTest` being modified here. Actually this method does not seem to be used. Only `getBlueprintUpgradeOptionForGA()` is called (from `BlueprintV4RequestToBlueprintConverter`).

**Review by lajosrodek — CHANGES_REQUESTED** [review]

**Review by lajosrodek — APPROVED** [review]
Thanks, LGTM

File: core/src/main/java/com/sequenceiq/cloudbreak/service/upgrade/UpgradePreconditionService.java
Comment (keyki) [inline core/src/main/java/com/sequenceiq/cloudbreak/service/upgrade/UpgradePreconditionService.java:62]: I don't think this is sufficient. This will make sure that the DL itself can be upgrade if there are Data Hubs with custom templates, but then the DH upgrade will still be rejected. If you take a look at this method, https://github.com/hortonworks/cloudbreak/blob/e0bd28ed3340d9e8ee3b2b9aa7315377b7e3371d/core-model/src/main/java/com/sequenceiq/cloudbreak/domain/BlueprintUpgradeOption.java#L48 I think this will still reject the upgrade for a Data hub with custom template.

File: core/src/main/java/com/sequenceiq/cloudbreak/service/upgrade/image/BlueprintUpgradeOptionValidator.java
Comment (nadamdb) [inline core/src/main/java/com/sequenceiq/cloudbreak/service/upgrade/image/BlueprintUpgradeOptionValidator.java:48]: Is this logic correct? Where this function is called on line #28, it is only called if the status is default, so the status will never be USER_MANAGED here.

**Review by nadamdb — CHANGES_REQUESTED** [review]

File: core/src/main/java/com/sequenceiq/cloudbreak/converter/v4/blueprint/BlueprintV4RequestToBlueprintConverter.java
Comment (adarshsharma1983) [inline core/src/main/java/com/sequenceiq/cloudbreak/converter/v4/blueprint/BlueprintV4RequestToBlueprintConverter.java:81]: Since we have started to set BlueprintUpgradeOption as part of this change, it should be null for existing records. Don't have any requirement to populate this column for existing records?

Comment (vivprabu) [reply core/src/main/java/com/sequenceiq/cloudbreak/converter/v4/blueprint/BlueprintV4RequestToBlueprintConverter.java:81]: there is no requirement to populate for all existing record Adarsh. UpgradePreConditionService: line 62 will take care of it

File: core/src/main/java/com/sequenceiq/cloudbreak/service/upgrade/UpgradePreconditionService.java
Comment (vivprabu) [reply core/src/main/java/com/sequenceiq/cloudbreak/service/upgrade/UpgradePreconditionService.java:62]: @keyki for standalone DH upgrade (DH upgrade without DL) the BluePrintUpgradeOption - is not being checked at all in code. @nadamdb pointed it out and I went through the code completely.

File: core/src/main/java/com/sequenceiq/cloudbreak/service/upgrade/image/BlueprintUpgradeOptionValidator.java
Comment (vivprabu) [reply core/src/main/java/com/sequenceiq/cloudbreak/service/upgrade/image/BlueprintUpgradeOptionValidator.java:48]: thank you for pointing it out @nadamdb will correct it.

**Review by nadamdb — CHANGES_REQUESTED** [review]
As discussed on today's meeting, this part:
https://github.com/hortonworks/cloudbreak/blob/255242734b03550d1d9975b3db2826090b0ba2b8/core/src/main/java/com/sequenceiq/distrox/v1/distrox/service/upgrade/DistroXUpgradeAvailabilityService.java#L153
should also be either
- fixed the same way like the other similar filtering is done (not taking the value into account if the blueprint is custom)
- fixed as above + adding an SQL migration script in a later version to change all custom blueprints to GA and remove this extra condition fix
- ad-hoc SQL after release to to change all custom blueprints to GA

**Comment by vivprabu** [comment]
> As discussed on today's meeting, this part:
> 
> https://github.com/hortonworks/cloudbreak/blob/255242734b03550d1d9975b3db2826090b0ba2b8/core/src/main/java/com/sequenceiq/distrox/v1/distrox/service/upgrade/DistroXUpgradeAvailabilityService.java#L153
> 
> 
> should also be either
> * fixed the same way like the other similar filtering is done (not taking the value into account if the blueprint is custom)
> * fixed as above + adding an SQL migration script in a later version to change all custom blueprints to GA and remove this extra condition fix
> * ad-hoc SQL after release to to change all custom blueprints to GA

done. please review again

File: core/src/main/java/com/sequenceiq/distrox/v1/distrox/service/upgrade/DistroXUpgradeAvailabilityService.java
Comment (nadamdb) [inline core/src/main/java/com/sequenceiq/distrox/v1/distrox/service/upgrade/DistroXUpgradeAvailabilityService.java:155]: Please add UT for the condition change.

File: core/src/main/java/com/sequenceiq/distrox/v1/distrox/service/upgrade/DistroXUpgradeAvailabilityService.java
Comment (nadamdb) [inline core/src/main/java/com/sequenceiq/distrox/v1/distrox/service/upgrade/DistroXUpgradeAvailabilityService.java:156]: nit: it would be nice to update the log line as well with something like "..., not GA and the blueprint is not a custom one".

**Review by nadamdb — CHANGES_REQUESTED** [review]

File: core/src/main/java/com/sequenceiq/distrox/v1/distrox/service/upgrade/DistroXUpgradeAvailabilityService.java
Comment (vivprabu) [reply core/src/main/java/com/sequenceiq/distrox/v1/distrox/service/upgrade/DistroXUpgradeAvailabilityService.java:155]: done. please review and approve

**Review by nadamdb — APPROVED** [review]
LGTM, thanks!

**Comment by keyki** [comment]
@vivprabu can you please rebase the code to the latest master and we can merge.

**Comment by vivprabu** [comment]
> @vivprabu can you please rebase the code to the latest master and we can merge.

@keyki done

**Pull request merged**
